src/iceberg/util/snapshot_util.cc

```diff
@@ -0,0 +1,373 @@
+ * under the License.
+ */
+
+#include "iceberg/schema.h"
+#include "iceberg/snapshot.h"
+#include "iceberg/table.h"
+#include "iceberg/table_metadata.h"
+#include "iceberg/util/snapshot_util_internal.h"
+#include "iceberg/util/timepoint.h"
+
+namespace iceberg {
+
+namespace {}  // namespace
+
+Result<std::vector<std::shared_ptr<Snapshot>>> SnapshotUtil::AncestorsOf(
+    const Table& table, int64_t snapshot_id) {
+  ICEBERG_ASSIGN_OR_RAISE(auto start, table.SnapshotById(snapshot_id));
+  if (!start) {
+    return InvalidArgument("Cannot find snapshot: {}", snapshot_id);
+  }
+  return AncestorsOf(table, start);
+}
+
+Result<bool> SnapshotUtil::IsAncestorOf(const Table& table, int64_t snapshot_id,
+                                        int64_t ancestor_snapshot_id) {
+  ICEBERG_ASSIGN_OR_RAISE(auto ancestors, AncestorsOf(table, snapshot_id));
+  for (const auto& snapshot : ancestors) {
+    if (snapshot->snapshot_id == ancestor_snapshot_id) {
+      return true;
+    }
+  }
+  return false;
+}
+
+Result<bool> SnapshotUtil::IsAncestorOf(const Table& table,
+                                        int64_t ancestor_snapshot_id) {
+  ICEBERG_ASSIGN_OR_RAISE(auto current, table.current_snapshot());
+  return IsAncestorOf(table, current->snapshot_id, ancestor_snapshot_id);
+}
+
+Result<bool> SnapshotUtil::IsParentAncestorOf(const Table& table, int64_t snapshot_id,
+                                              int64_t ancestor_parent_snapshot_id) {
+  ICEBERG_ASSIGN_OR_RAISE(auto ancestors, AncestorsOf(table, snapshot_id));
+  for (const auto& snapshot : ancestors) {
+    if (snapshot->parent_snapshot_id.has_value() &&
+        snapshot->parent_snapshot_id.value() == ancestor_parent_snapshot_id) {
+      return true;
+    }
+  }
+  return false;
+}
+
+Result<std::vector<std::shared_ptr<Snapshot>>> SnapshotUtil::CurrentAncestors(
+    const Table& table) {
+  ICEBERG_ASSIGN_OR_RAISE(auto current, table.current_snapshot());
+  return AncestorsOf(table, current);
+}
+
+Result<std::vector<int64_t>> SnapshotUtil::CurrentAncestorIds(const Table& table) {
+  ICEBERG_ASSIGN_OR_RAISE(auto ancestors, CurrentAncestors(table));
+  return ToIds(ancestors);
+}
+
+Result<std::optional<std::shared_ptr<Snapshot>>> SnapshotUtil::OldestAncestor(
+    const Table& table) {
+  ICEBERG_ASSIGN_OR_RAISE(auto ancestors, CurrentAncestors(table));
+  if (ancestors.empty()) {
+    return std::nullopt;
+  }
+  return ancestors.back();
+}
+
+Result<std::optional<std::shared_ptr<Snapshot>>> SnapshotUtil::OldestAncestorOf(
+    const Table& table, int64_t snapshot_id) {
+  ICEBERG_ASSIGN_OR_RAISE(auto ancestors, AncestorsOf(table, snapshot_id));
+  if (ancestors.empty()) {
+    return std::nullopt;
+  }
+  return ancestors.back();
+}
+
+Result<std::optional<std::shared_ptr<Snapshot>>> SnapshotUtil::OldestAncestorAfter(
+    const Table& table, TimePointMs timestamp_ms) {
+  ICEBERG_ASSIGN_OR_RAISE(auto current, table.current_snapshot());
```

Comment:
Actually it is not an error if current snapshot is unavailable. We can return std::nullopt when the result error is `NotFound`.